server/statistics/util.go

```diff
@@ -53,3 +56,274 @@ func (r *RollingStats) Median() float64 {
+	Less(then TopNItem) bool
+}
+
+// indexedHeap is a heap with index.
+type indexedHeap struct {
+	rev   bool
+	items []TopNItem
+	index map[uint64]int
+}
+
+func newTopNHeap() *indexedHeap {
+	return &indexedHeap{
+		rev:   false,
+		items: []TopNItem{},
+		index: make(map[uint64]int),
+	}
+}
+
+func newRevTopNHeap() *indexedHeap {
+	return &indexedHeap{
+		rev:   true,
+		items: []TopNItem{},
+		index: make(map[uint64]int),
+	}
+}
+
+// Implementing heap.Interface.
+func (hp *indexedHeap) Len() int {
+	return len(hp.items)
+}
+
+// Implementing heap.Interface.
+func (hp *indexedHeap) Less(i, j int) bool {
+	if !hp.rev {
+		return hp.items[i].Less(hp.items[j])
+	}
+	return hp.items[j].Less(hp.items[i])
+}
+
+// Implementing heap.Interface.
+func (hp *indexedHeap) Swap(i, j int) {
+	lid := hp.items[i].ID()
+	rid := hp.items[j].ID()
+	hp.items[i], hp.items[j] = hp.items[j], hp.items[i]
+	hp.index[lid] = j
+	hp.index[rid] = i
+}
+
+// Implementing heap.Interface.
+func (hp *indexedHeap) Push(x interface{}) {
+	item := x.(TopNItem)
+	hp.index[item.ID()] = hp.Len()
+	hp.items = append(hp.items, item)
+}
+
+// Implementing heap.Interface.
+func (hp *indexedHeap) Pop() interface{} {
+	l := hp.Len()
+	item := hp.items[l-1]
+	hp.items = hp.items[:l-1]
+	delete(hp.index, item.ID())
+	return item
+}
+
+// Top returns the top item.
+func (hp *indexedHeap) Top() TopNItem {
+	if hp.Len() <= 0 {
+		return nil
+	}
+	return hp.items[0]
+}
+
+// Get returns item with the given ID.
+func (hp *indexedHeap) Get(id uint64) TopNItem {
+	idx, ok := hp.index[id]
+	if !ok {
+		return nil
+	}
+	item := hp.items[idx]
+	return item.(TopNItem)
+}
+
+// GetAll returns all the items.
+//
+// Callers should not mutate the returned slice.
+func (hp *indexedHeap) GetAll() []TopNItem {
+	return hp.items
+}
+
+// Put inserts item or updates the old item if it exists.
+func (hp *indexedHeap) Put(item TopNItem) (isNew bool) {
+	idx, ok := hp.index[item.ID()]
+	if !ok {
+		heap.Push(hp, item)
+	} else {
+		hp.items[idx] = item
+		heap.Fix(hp, idx)
+	}
+	return !ok
+}
+
+// Fix fixes the heap, returns false if there is no item has the given ID.
+func (hp *indexedHeap) Fix(id uint64) (ok bool) {
+	idx, ok := hp.index[id]
+	if ok {
+		heap.Fix(hp, idx)
+	}
+	return
+}
+
+// Remove deletes item by ID and returns it.
+func (hp *indexedHeap) Remove(id uint64) TopNItem {
+	idx, ok := hp.index[id]
+	if !ok {
+		return nil
+	}
+	item := heap.Remove(hp, idx)
+	return item.(TopNItem)
+}
+
+func maxInt(x int, y int) int {
+	if x > y {
+		return x
+	}
+	return y
+}
+
+// TopN maintains the N largest items.
+// NOTE: TopN is thread-safe.
+type TopN struct {
+	rw   sync.RWMutex
+	n    int
+	topn *indexedHeap
+	rest *indexedHeap
+}
+
+// NewTopN returns a TopN working on the items.
+func NewTopN(n int) *TopN {
+	return &TopN{
+		n:    maxInt(n, 1),
+		topn: newTopNHeap(),
+		rest: newRevTopNHeap(),
+	}
+}
+
+func (tn *TopN) lock() {
```

Comment:
you can use 

```go
type TopN struct {
    sync.RwMutex
}
```

so we don't need to provide these lock functions.

Reply:
I don't want to export these lock functions.

Reply:
please remove these, too verbose. 

to follow the lock style in tidb, you can use 

```go
type TopN struct {
     mu        struct {
		sync.RWMutex
		// fields you want to protect
	}
}
```

Reply:
Removed, but not the tidb lock style.